### ISSUE TITLE
V100

### ISSUE THREAD

**Issue by gaohao-dev** [HazyResearch/flash-attention#148]
V100  GPU  not supported

**Comment by tridao** [comment]
That's right, as mentioned in the README, we support Turing, Ampere, Ada, or Hopper GPUs (e.g., H100, A100, RTX 3090, T4, RTX 2080). There's plan to support V100 in June.

**Comment by dimaischenko** [comment]
> That's right, as mentioned in the README, we support Turing, Ampere, Ada, or Hopper GPUs (e.g., H100, A100, RTX 3090, T4, RTX 2080). There's plan to support V100 in June.

Support for V100 will be very cool!

**Comment by pandazki** [comment]
> 

Looking forward to support for V100.

**Comment by tridao** [comment]
I'm no longer planning to work on V100, but we welcome community's contribution.

**Comment by tridao** [comment]
The tensor cores (for matmul) on Voltas are different compared to the tensor cores on Turing and Ampere, and the shared memory layout required is different.

Each Volta tensor core multiplies 4 matrices of shape 8x4 and 4x4 (i.e. the mma.sync.aligned.m8n8k4 instruction). Each Turing and Ampere tensor core multiply 1 matrix of shape 16x8 and 8x8, or 16x16 and 16x8 (i.e. the mma.sync.aligned.m16n8k8 and mma.sync.aligned.m16n8k16 instruction). This means that inputs to the mma instructions need to be laid out differently in shared memory. This shared memory load/store is the trickiest part.

**Comment by tridao** [comment]
To use tensor cores we need to call the mma instructions, either directly [here](https://github.com/HazyResearch/flash-attention/blob/9818f85fee29ac6b60c9214bce841f8109a18b1b/csrc/flash_attn/src/fmha/gemm.h#L189), or by calling [cutlass](https://github.com/HazyResearch/flash-attention/blob/9818f85fee29ac6b60c9214bce841f8109a18b1b/csrc/flash_attn/src/fmha/gemm.h#L275) (which then calls the mma instructions).

**Comment by calebthomas259** [comment]
Hi @tridao, could you please briefly explain the reason for using 2 buffers for some shared memory tiles? e.g. [`Smem_tile_q`](https://github.com/HazyResearch/flash-attention/blob/01c40dacc45c91f80b74d4a29f91a44eddd64011/csrc/flash_attn/src/fmha/kernel_traits.h#L54) uses two buffers, and you also mention it when explaining the shared memory layout [here](https://github.com/HazyResearch/flash-attention/issues/190#issuecomment-1546368919).

Also, this is probably related, but I don't understand what the "boundary" means in [Smem_tile_without_skews](https://github.com/HazyResearch/flash-attention/blob/01c40dacc45c91f80b74d4a29f91a44eddd64011/csrc/flash_attn/src/fmha/smem_tile.h#L108)... do you have any hints for this ahaha

Thanks!

**Comment by AmgadHasan** [comment]
Hi @tridao !
Any update regarding the V100?
Does it currently support flash attention 1 and flash attention 2?

Thanks

**Comment by tridao** [comment]
We have no plan to support V100.

**Comment by nahidalam** [comment]
For V100, it is better to use `xformers` library. I was able to pretrain LLaVA on V100 using `xformers`
<img width="365" alt="Screenshot 2024-04-17 at 9 26 52 PM" src="https://github.com/Dao-AILab/flash-attention/assets/2341472/5ed5870b-965f-4101-87c6-a54bc5fcb0d7">

**Comment by ai-bond** [comment]
> Support for V100 will be very cool!

I've tried to done it with Volta smem and mma shapes/precision/func restrictions-> [Flash attention 2 for Volta](https://github.com/ai-bond/flash-attention-v100)
